README.md
# language-asn

[![Hackage](https://img.shields.io/hackage/v/language-asn.svg)](https://hackage.haskell.org/package/language-asn)
[![BSD3 license](https://img.shields.io/badge/license-BSD3-blue.svg)](LICENSE)
[![Stackage Lts](http://stackage.org/package/language-asn/badge/lts)](http://stackage.org/lts/package/language-asn)
[![Stackage Nightly](http://stackage.org/package/language-asn/badge/nightly)](http://stackage.org/nightly/package/language-asn)

asn.1 encoding and decoding

CHANGELOG.md
# Changelog

`language-asn` uses [PVP Versioning][1].
The changelog is available [on GitHub][2].

0.1.1.0
=======

* Added `FromJSON`, `FromJSONKey`, and `ToJSON` instances to
  `ObjectIdentifier`.

0.1.0.0
=======

* Moved all ASN.1 code from the now-deprecated `asn1-codec` into
  this library.

[1]: https://pvp.haskell.org
[2]: https://github.com/chessai/language-asn/releases


src/Language/Asn/Decoding.hs
{-# language ExistentialQuantification #-}
{-# language GADTs #-}
{-# language OverloadedStrings #-}
{-# language GeneralizedNewtypeDeriving #-}
{-# language BangPatterns #-}
{-# language MagicHash #-}

module Language.Asn.Decoding
  ( ber
  , sequence
  , sequenceOf
  , required
  , optional
  , defaulted
  , utf8String
  , integer
  , integerRanged
  , int32
  , int
  , word32
  , word64
  , null
  , null'
  , octetString
  , octetStringWord8
  , octetStringWord32
  , objectIdentifier
  , choice
  , option
  , tag
  , mapFailable
  ) where

import Prelude hiding (sequence,null)
import Language.Asn.Types.Internal
import Data.ByteString (ByteString)
import Data.Bits (Bits(..))
import Control.Monad hiding (sequence)
import Data.Maybe (fromMaybe, listToMaybe)
import Data.Text (Text)
import Data.Word (Word8, Word32, Word64)
import Data.Int (Int32)
import Data.Functor.Identity (Identity(..))
import Text.Printf (printf)
import qualified GHC.Exts as E
import qualified Data.Text.Encoding as TE
import qualified Data.List as List
import qualified Data.ByteString as ByteString
import qualified Data.ByteString.Unsafe as BSU

ber :: AsnDecoding a -> ByteString -> Either String a
ber d bs = requireNoLeftovers =<< decodeBerInternal d Nothing bs

sequence :: FieldDecoding a -> AsnDecoding a
sequence = AsnDecodingSequence

sequenceOf :: AsnDecoding a -> AsnDecoding [a]
sequenceOf = AsnDecodingSequenceOf id

required :: FieldName -> AsnDecoding a -> FieldDecoding a
required name d = FieldDecoding (liftAp (FieldDecodingRequired name d))

optional :: FieldName -> AsnDecoding a -> FieldDecoding (Maybe a)
optional name d = FieldDecoding (liftAp (FieldDecodingOptional name d id))

defaulted :: Show a => FieldName -> AsnDecoding a -> a -> FieldDecoding a
defaulted name d defVal = FieldDecoding (liftAp (FieldDecodingDefault name d defVal show))

choice :: [OptionDecoding a] -> AsnDecoding a
choice = AsnDecodingChoice

option :: OptionName -> AsnDecoding a -> OptionDecoding a
option = OptionDecoding

tag :: TagClass -> Int -> Explicitness -> AsnDecoding a -> AsnDecoding a
tag c n e = AsnDecodingRetag (TagAndExplicitness (Tag c n) e)

utf8String :: AsnDecoding Text
utf8String = AsnDecodingUniversal $ UniverseDecodingTextualString Utf8String id (Subtypes []) (Subtypes [])

octetString :: AsnDecoding ByteString
octetString = AsnDecodingUniversal $ UniverseDecodingOctetString id (Subtypes [])

octetStringWord32 :: AsnDecoding Word32
octetStringWord32 = mapFailable
  ( \bs -> if ByteString.length bs == 4
      then Right $ (fromIntegral :: Word -> Word32)
         $ unsafeShiftL 24 (fromIntegral (BSU.unsafeIndex bs 0))
         + unsafeShiftL 16 (fromIntegral (BSU.unsafeIndex bs 1))
         + unsafeShiftL 8 (fromIntegral (BSU.unsafeIndex bs 2))
         + fromIntegral (BSU.unsafeIndex bs 3)
      else Left "octetStringWord32 expects the octet string to have exactly 4 bytes"
  ) octetString

octetStringWord8 :: AsnDecoding Word8
octetStringWord8 = mapFailable
  ( \bs -> if ByteString.length bs == 1
      then Right (BSU.unsafeIndex bs 0)
      else Left "octetStringWord8 expects the octet string to have exactly 1 byte"
  ) octetString

integer :: AsnDecoding Integer
integer = AsnDecodingUniversal (UniverseDecodingInteger id (Subtypes []))

-- This could be improved by making sure that integer in question is actually
-- in the provided bounds.
int :: AsnDecoding Int
int = AsnDecodingUniversal (UniverseDecodingInteger fromIntegral (Subtypes []))

-- This could be improved by making sure that integer in question is actually
-- in the provided bounds.
int32 :: AsnDecoding Int32
int32 = AsnDecodingUniversal (UniverseDecodingInteger fromIntegral (Subtypes []))

-- This could be improved by making sure that integer in question is actually
-- in the provided bounds.
word32 :: AsnDecoding Word32
word32 = AsnDecodingUniversal (UniverseDecodingInteger fromIntegral (Subtypes []))

-- This could be improved by making sure that integer in question is actually
-- in the provided bounds.
word64 :: AsnDecoding Word64
word64 = AsnDecodingUniversal (UniverseDecodingInteger fromIntegral (Subtypes []))

null :: AsnDecoding ()
null = AsnDecodingUniversal (UniverseDecodingNull ())

null' :: a -> AsnDecoding a
null' = AsnDecodingUniversal . UniverseDecodingNull

objectIdentifier :: AsnDecoding ObjectIdentifier
objectIdentifier = AsnDecodingUniversal (UniverseDecodingObjectIdentifier id (Subtypes []))

integerRanged :: Integer -> Integer -> AsnDecoding Integer
integerRanged lo hi = AsnDecodingUniversal
  (UniverseDecodingInteger id (Subtypes [SubtypeValueRange lo hi]))

mapFailable :: (a -> Either String b) -> AsnDecoding a -> AsnDecoding b
mapFailable f d = AsnDecodingConversion d f

repeatUntilEmpty :: Monad m => (ByteString -> m (a,ByteString)) -> ByteString -> m [a]
repeatUntilEmpty f = go where
  go bs1 = do
    (a,bs2) <- f bs1
    if ByteString.null bs2
      then return [a]
      else fmap (a:) (go bs2)

decodeBerInternal :: AsnDecoding a -> Maybe Tag -> ByteString -> Either String (a,ByteString)
decodeBerInternal x overrideTag bs1 = case x of
  AsnDecodingUniversal u -> do
    (bsContent,bsRemainder) <- takeTagAndLength Primitive (Tag Universal (universeDecodingTagNumber u))
    a <- decodeUniversal u bsContent
    return (a,bsRemainder)
  AsnDecodingRetag (TagAndExplicitness newTag expl) nextDecoding -> case expl of
    Implicit -> decodeBerInternal nextDecoding (Just $ fromMaybe newTag overrideTag) bs1
    Explicit -> do
      (bsContent,bsRemainder) <- takeTagAndLength Constructed newTag
      a <- requireNoLeftovers =<< decodeBerInternal nextDecoding Nothing bsContent
      return (a,bsRemainder)
  AsnDecodingSequence (FieldDecoding fieldDecoding) -> do
    (bsContent,bsRemainder) <- takeTagAndLength Constructed sequenceTag
    a <- requireNoLeftovers =<< getDecodePart (runAp decodeField fieldDecoding) bsContent
    return (a,bsRemainder)
  AsnDecodingSequenceOf f nextDecoding -> do
    (bsContent,bsRemainder) <- takeTagAndLength Constructed sequenceTag
    cs <- repeatUntilEmpty (decodeBerInternal nextDecoding Nothing) bsContent
    return (f cs,bsRemainder)
  AsnDecodingConversion nextDecoding conv -> do
    (b,bs2) <- decodeBerInternal nextDecoding overrideTag bs1
    a <- conv b
    return (a,bs2)
  -- Note: overrideTag is currently ignored in this case. Per
  -- ASN.1 rules, you cannot put an IMPLICIT tag over a CHOICE,
  -- which is the only thing that could cause this to happen.
  -- Still, I don't like that it's being ignored.
  AsnDecodingChoice opts -> case ByteString.uncons bs1 of
    Nothing -> Left "while trying to decode Choice tag, ran out of input"
    Just (b,bs2) -> do
      let possibilities = nextExpectedTags =<< map (\(OptionDecoding _ d) -> d) opts
          theTagNumberPrefix = b .&. 31
      (theTagNumber,_) <- decipherTagNumber theTagNumberPrefix bs2
      let mmatched = listToMaybe $ List.filter
            (\((Tag tc tn,cstn),_) -> b .&. 192 == tagClassBit tc && b .&. 32 == constructionBit cstn && tn == theTagNumber)
            possibilities
      case mmatched of
        Nothing -> Left $ concat 
          [ "while trying to decode Choice tag, the tag did not match any of the expected tags: found "
          , show theTagNumber
          , " but expected one of ["
          , List.intercalate "," $ flip map possibilities $ \((Tag tc tn, cstn),_) -> concat
              [ show tc
              , " "
              , show cstn
              , " "
              , show tn
              ]
          , "]"
          ]
        Just (_,Wrapper chosenDecoder conv2) -> do
          (c,bs3) <- decodeBerInternal chosenDecoder Nothing bs1
          r <- conv2 c
          Right (r,bs3)
  where
  takeTagAndLength :: Construction -> Tag -> Either String (ByteString,ByteString)
  takeTagAndLength construction t1 = do
    let tagToUse = fromMaybe t1 overrideTag
    bs2 <- expectTag construction tagToUse bs1
    (len,bs3) <- takeLength bs2
    splitOrFail len bs3

data Wrapper a = forall b. Wrapper (AsnDecoding b) (b -> Either String a)

idWrapper :: AsnDecoding a -> Wrapper a
idWrapper d = Wrapper d Right

nextExpectedTags :: AsnDecoding a -> [((Tag,Construction), Wrapper a)]
nextExpectedTags x = case x of
  AsnDecodingUniversal u -> [((Tag Universal (universeDecodingTagNumber u), Primitive),idWrapper x)]
  AsnDecodingSequenceOf _ _ -> [((sequenceTag, Constructed),idWrapper x)]
  AsnDecodingSequence _ -> [((sequenceTag, Constructed),idWrapper x)]
  AsnDecodingChoice opts -> join (map (\(OptionDecoding _ theDec) -> nextExpectedTags theDec) opts)
  AsnDecodingRetag (TagAndExplicitness newTag expl) nextDecoding -> case expl of
    Explicit -> [((newTag,Constructed),idWrapper x)]
    Implicit -> map (\((_,c),Wrapper theDec theConv) -> ((newTag,c), Wrapper (AsnDecodingRetag (TagAndExplicitness newTag expl) theDec) theConv)) (nextExpectedTags nextDecoding)
  AsnDecodingConversion nextDecoding conv ->
    map (\((t,c),Wrapper theDec theConv) -> ((t,c),Wrapper theDec (theConv >=> conv))) (nextExpectedTags nextDecoding)

decodeField :: FieldDecodingPart a -> DecodePart a
decodeField x = case x of
  FieldDecodingRequired _ d -> DecodePart (decodeBerInternal d Nothing)
  FieldDecodingOptional _ d conv1 -> handlePossiblyMissingField d conv1
  FieldDecodingDefault _ d a _ -> handlePossiblyMissingField d (fromMaybe a)

handlePossiblyMissingField :: AsnDecoding b -> (Maybe b -> a) -> DecodePart a
handlePossiblyMissingField d conv1 = DecodePart $ \bs1 -> case ByteString.uncons bs1 of
  Nothing -> Right (conv1 Nothing, bs1)
  Just (b,bs2) -> do
    let possibilities = nextExpectedTags d
        theTagNumberPrefix = b .&. 31
    (theTagNumber,_) <- decipherTagNumber theTagNumberPrefix bs2
    let mmatched = listToMaybe $ List.filter
          (\((Tag tc tn,cstn),_) -> b .&. 192 == tagClassBit tc && b .&. 32 == constructionBit cstn && tn == theTagNumber) possibilities
    case mmatched of
      Nothing -> Right (conv1 Nothing, bs1)
      Just (_,Wrapper chosenDecoder conv2) -> do
        (c,bs3) <- decodeBerInternal chosenDecoder Nothing bs1
        r <- conv2 c
        Right (conv1 (Just r),bs3)

-- The first argument is the full tag byte
decipherTagNumber :: Word8 -> ByteString -> Either String (Int,ByteString)
decipherTagNumber w bs = do
  let tn = w .&. 31
  if tn < 31
    then Right (fromIntegral tn,bs)
    else error "decipherTagNumber: handle tags greater than 30"


requireNoLeftovers :: (a,ByteString) -> Either String a
requireNoLeftovers (a,bs) = if ByteString.null bs then Right a else Left "expected not to have leftovers, but there were leftovers"

expectTag :: Construction -> Tag -> ByteString -> Either String ByteString
expectTag construction (Tag tc tn) bs = case ByteString.uncons bs of
  Nothing -> Left "expected a byte for the Tag but the ByteString was empty"
  Just (b,bsNext) -> do
    let expectedTagBits = tagClassBit tc
        actualTagBits = b .&. 192
    when (actualTagBits /= expectedTagBits)
      $ Left $ "while parsing the tag, the tag class bits did not match: "
            ++ "found " ++ printf "%08b" actualTagBits ++ " but expected "
            ++ printf "%08b" expectedTagBits ++ " (" ++ show tc ++ ")"
    when (b .&. 32 /= constructionBit construction) $ Left "while parsing the tag, the construction bit did not match"
    if tn < 31
      then do
        when (b .&. 31 /= fromIntegral tn) $ Left "while parsing the tag, the tag number did not match what was expected"
        Right bsNext
      else error "expectTag: handle tag numbers higher than 30"

splitOrFail :: Int -> ByteString -> Either String (ByteString,ByteString)
splitOrFail i bs = if i > ByteString.length bs
  then Left "tried to take a fixed number of bytes as specified by the encoding, but bytestring ended"
  else Right (ByteString.splitAt i bs)

takeLength :: ByteString -> Either String (Int,ByteString)
takeLength bs1 = case ByteString.uncons bs1 of
  Nothing -> Left "while trying to decode the length, expected an initial octet but the bytestring ended"
  Just (b,bs2) -> if b < 128
    then Right (fromIntegral b, bs2)
    else let bytesToTake = fromIntegral (127 .&. b) in
      if ByteString.length bs2 < bytesToTake
        then Left "while decoding multi-byte length, ran out of bytes"
        else let (bs3,bs4) = ByteString.splitAt bytesToTake bs2 in Right (parseLength bs3, bs4)


parseLength :: ByteString -> Int
parseLength = ByteString.foldl' (\i w8 -> i * 256 + fromIntegral w8) 0

-- This currently does the wrong thing for negative numbers
parseInteger :: ByteString -> Integer
parseInteger = ByteString.foldl' (\i w8 -> i * 256 + fromIntegral w8) 0

decodeUniversal :: UniverseDecoding a -> ByteString -> Either String a
decodeUniversal x bs = case x of
  UniverseDecodingInteger f _ -> Right (f (parseInteger bs))
  UniverseDecodingTextualString _ f _ _ -> case TE.decodeUtf8' bs of
    Left _ -> Left "while decoding string primitive, found that UTF8-encoding was not used"
    Right t -> Right (f t)
  UniverseDecodingObjectIdentifier f _ -> fmap f (stepOidAll bs)
  UniverseDecodingOctetString f _ -> Right (f bs)
  UniverseDecodingNull a -> Right a

stepOidAll :: ByteString -> Either String ObjectIdentifier
stepOidAll bs1 = case ByteString.uncons bs1 of
  Nothing -> Left "while decoding OID, found no bytes, the OID should have at least one octet"
  Just (b,bs2) ->
    let (w1,w2) = quotRem b 40
        Identity nums = repeatUntilEmpty (Identity . stepOid 0) bs2
     in Right (ObjectIdentifier (E.fromList $ fromIntegral w1 : fromIntegral w2 : nums))

stepOid :: Word -> ByteString -> (Word,ByteString)
stepOid !i bs1 = case ByteString.uncons bs1 of
  Nothing -> (i,bs1)
  Just (w,bs2) ->
    let !acc = i * 128 + fromIntegral (w .&. 127) in
    if w .&. 128 == 128
      then stepOid acc bs2
      else (acc,bs2)

universeDecodingTagNumber :: UniverseDecoding a -> Int
universeDecodingTagNumber x = case x of
  UniverseDecodingInteger _ _ -> 2
  UniverseDecodingTextualString typ _ _ _ -> tagNumStringType typ
  UniverseDecodingObjectIdentifier _ _ -> 6
  UniverseDecodingOctetString _ _ -> 4
  UniverseDecodingNull _ -> 5





src/Language/Asn/Types/Internal.hs
{-# language BangPatterns #-}
{-# language CPP #-}
{-# language DeriveFunctor #-}
{-# language DeriveGeneric #-}
{-# language ExistentialQuantification #-}
{-# language GADTs #-}
{-# language GeneralizedNewtypeDeriving #-}
{-# language LambdaCase #-}
{-# language MagicHash #-}
{-# language OverloadedStrings #-}
{-# language RankNTypes #-}
{-# language StandaloneDeriving #-}

module Language.Asn.Types.Internal where

import Prelude hiding (sequence,null)

import Data.Aeson (ToJSON,FromJSON,FromJSONKey)
import Data.ByteString (ByteString)
import Data.Functor.Contravariant (Contravariant(..))
import Data.Hashable (Hashable(..))
import Data.Primitive (PrimArray)
import Data.Semigroup (Semigroup)
import Data.String (IsString)
import Data.Text (Text)
import Data.Word (Word64,Word8)
import GHC.Generics (Generic)
import GHC.Int (Int(..))

import qualified Data.Aeson as AE
import qualified Data.Aeson.Types as AE
import qualified Data.ByteString.Lazy as LB
import qualified Data.List as L
import qualified Data.Text as T
import qualified Data.Text.Read as TR
import qualified GHC.Exts as E

#if !MIN_VERSION_base(4,11,0)
import Data.Monoid (Monoid)
#endif

data AsnEncoding a
  = EncSequence [Field a]
  | forall b. EncSequenceOf (a -> [b]) (AsnEncoding b)
  | EncChoice (Choice a)
  | EncRetag TagAndExplicitness (AsnEncoding a)
  | EncUniversalValue (UniversalValue a)

instance Contravariant AsnEncoding where
  contramap f = \case
    EncRetag te y -> EncRetag te (contramap f y)
    EncUniversalValue u -> EncUniversalValue (contramap f u)
    EncSequence xs -> EncSequence (map (contramap f) xs)
    EncChoice c -> EncChoice (contramap f c)
    EncSequenceOf conv enc -> EncSequenceOf (conv . f) enc

data UniversalValue a
  = UniversalValueBoolean (a -> Bool) (Subtypes Bool)
  | UniversalValueInteger (a -> Integer) (Subtypes Integer)
  | UniversalValueNull
  | UniversalValueOctetString (a -> ByteString) (Subtypes ByteString)
  | UniversalValueTextualString StringType (a -> Text) (Subtypes Text) (Subtypes Char)
  | UniversalValueObjectIdentifier (a -> ObjectIdentifier) (Subtypes ObjectIdentifier)

instance Contravariant UniversalValue where
  contramap f = \case
    UniversalValueBoolean conv s -> UniversalValueBoolean (conv . f) s
    UniversalValueInteger conv s -> UniversalValueInteger (conv . f) s
    UniversalValueObjectIdentifier conv s -> UniversalValueObjectIdentifier (conv . f) s
    UniversalValueOctetString conv s -> UniversalValueOctetString (conv . f) s
    UniversalValueTextualString typ conv s1 s2 -> UniversalValueTextualString typ (conv . f) s1 s2
    UniversalValueNull -> UniversalValueNull

newtype Subtypes a = Subtypes { getSubtypes :: [Subtype a] }
  deriving (Semigroup,Monoid)

-- | Note: we deviate slightly from the actual definition of an object
-- identifier. Technically, each number of an OID should be allowed to
-- be an integer of unlimited size. However, we are intentionally unfaithful
-- to this specification because in practice, there are no OIDs that use
-- integers above a 32-bit word, so we just use the machine's native word
-- size.
newtype ObjectIdentifier = ObjectIdentifier
  { getObjectIdentifier :: PrimArray Word
  } deriving (Eq,Ord,Show,Generic)

instance ToJSON ObjectIdentifier where
  toJSON (ObjectIdentifier xs) = AE.toJSON
    $ mconcat
    $ L.intersperse (T.singleton '.')
    $ map (T.pack . show)
    $ E.toList xs

instance FromJSONKey ObjectIdentifier where
  fromJSONKey = AE.FromJSONKeyTextParser oidJsonParser

instance FromJSON ObjectIdentifier where
  parseJSON = AE.withText "ObjectIdentifier" oidJsonParser

oidJsonParser :: Text -> AE.Parser ObjectIdentifier
oidJsonParser t =
  case fmap (ObjectIdentifier . E.fromList) (mapM parseWord (T.splitOn (T.singleton '.') t)) of
    Nothing -> fail ("could not recognize " ++ T.unpack t ++ " as an OID")
    Just x -> return x

parseWord :: Text -> Maybe Word
parseWord t = case TR.decimal t of
  Right (i,leftovers) -> if T.null leftovers then Just i else Nothing
  Left _ -> Nothing

instance Hashable ObjectIdentifier where
  hash (ObjectIdentifier v) = hash (E.toList v)
  hashWithSalt s (ObjectIdentifier v) = hashWithSalt s (E.toList v)

newtype ObjectIdentifierSuffix = ObjectIdentifierSuffix
  { getObjectIdentifierSuffix :: PrimArray Word
  } deriving (Eq,Ord,Show,Generic)

instance Hashable ObjectIdentifierSuffix where
  hash (ObjectIdentifierSuffix v) = hash (E.toList v)
  hashWithSalt s (ObjectIdentifierSuffix v) = hashWithSalt s (E.toList v)

data Subtype a
  = SubtypeSingleValue a -- This also acts as PermittedAlphabet
  | SubtypeValueRange a a

data StringType
  = Utf8String
  | NumericString
  | PrintableString
  | TeletexString
  | VideotexString
  | IA5String
  | GraphicString
  | VisibleString
  | GeneralString
  | UniversalString
  | CharacterString
  | BmpString

data Explicitness = Explicit | Implicit
data TagAndExplicitness = TagAndExplicitness Tag Explicitness

data Choice a = forall b. Choice (a -> b) [b] (b -> ValueAndEncoding)

instance Contravariant Choice where
  contramap f (Choice conv bs bToValEnc) =
    Choice (conv . f) bs bToValEnc

data ValueAndEncoding = forall b. ValueAndEncoding Int OptionName b (AsnEncoding b)
data Field a
  = forall b. FieldRequired FieldName (a -> b) (AsnEncoding b)
  | forall b. FieldOptional FieldName (a -> Maybe b) (AsnEncoding b)
  | forall b. FieldDefaulted FieldName (a -> b) b (b -> String) (b -> b -> Bool) (AsnEncoding b)

instance Contravariant Field where
  contramap f = \case
    FieldRequired name g enc -> FieldRequired name (g . f) enc
    FieldOptional name g enc -> FieldOptional name (g . f) enc
    FieldDefaulted name g b1 b2 b3 enc -> FieldDefaulted name (g . f) b1 b2 b3 enc

data TaggedByteString = TaggedByteString !Construction !Tag !LB.ByteString
data TaggedStrictByteString = TaggedStrictByteString !Construction !Tag !ByteString
data Construction = Constructed | Primitive
  deriving (Show,Eq)

newtype FieldName = FieldName { getFieldName :: String }
  deriving (IsString)
newtype OptionName = OptionName { getOptionName :: String }
  deriving (IsString)

data TagClass
  = Universal
  | Application
  | Private
  | ContextSpecific
  deriving (Show,Eq)

data Tag = Tag
  { tagClass :: TagClass
  , tagNumber :: Int
  } deriving (Show,Eq)

fromIntegerTagAndExplicitness :: Integer -> TagAndExplicitness
fromIntegerTagAndExplicitness n = TagAndExplicitness
  (Tag ContextSpecific (fromIntegral n))
  Explicit

fromIntegerTag :: Integer -> Tag
fromIntegerTag n = Tag ContextSpecific (fromIntegral n)

------------------------------
-- Stuff specific to decoding
------------------------------

data AsnDecoding a
  = AsnDecodingUniversal (UniverseDecoding a)
  | forall b. AsnDecodingSequenceOf ([b] -> a) (AsnDecoding b)
  | forall b. AsnDecodingConversion (AsnDecoding b) (b -> Either String a)
  | AsnDecodingRetag TagAndExplicitness (AsnDecoding a)
  | AsnDecodingSequence (FieldDecoding a)
  | AsnDecodingChoice [OptionDecoding a]

deriving instance Functor AsnDecoding

data Ap f a where
  Pure :: a -> Ap f a
  Ap :: f a -> Ap f (a -> b) -> Ap f b

instance Functor (Ap f) where
  fmap f (Pure a)   = Pure (f a)
  fmap f (Ap x y)   = Ap x ((f .) <$> y)

instance Applicative (Ap f) where
  pure = Pure
  Pure f <*> y = fmap f y
  Ap x y <*> z = Ap x (flip <$> y <*> z)

data OptionDecoding a = OptionDecoding OptionName (AsnDecoding a)
  deriving (Functor)

newtype FieldDecoding a = FieldDecoding (Ap FieldDecodingPart a)
  deriving (Functor,Applicative)

data FieldDecodingPart a
  = FieldDecodingRequired FieldName (AsnDecoding a)
  | FieldDecodingDefault FieldName (AsnDecoding a) a (a -> String)
  | forall b. FieldDecodingOptional FieldName (AsnDecoding b) (Maybe b -> a)

data UniverseDecoding a
  = UniverseDecodingInteger (Integer -> a) (Subtypes Integer)
  | UniverseDecodingTextualString StringType (Text -> a) (Subtypes Text) (Subtypes Char)
  | UniverseDecodingOctetString (ByteString -> a) (Subtypes ByteString)
  | UniverseDecodingObjectIdentifier (ObjectIdentifier -> a) (Subtypes ObjectIdentifier)
  | UniverseDecodingNull a
  deriving (Functor)

newtype DecodePart a = DecodePart { getDecodePart :: ByteString -> Either String (a,ByteString) }
  deriving (Functor)

instance Applicative DecodePart where
  pure a = DecodePart (\bs -> Right (a,bs))
  DecodePart f <*> DecodePart g = DecodePart $ \bs1 -> do
    (h,bs2) <- f bs1
    (a,bs3) <- g bs2
    return (h a, bs3)

runAp :: Applicative g => (forall x. f x -> g x) -> Ap f a -> g a
runAp _ (Pure x) = pure x
runAp u (Ap f x) = flip id <$> u f <*> runAp u x

liftAp :: f a -> Ap f a
liftAp x = Ap x (Pure id)
{-# INLINE liftAp #-}

--------------------------
-- Functions common to encoding and decoding
--------------------------

-- Bit six is 1 when a value is constructed.
constructionBit :: Construction -> Word8
constructionBit = \case
  Constructed -> 32
  Primitive -> 0

-- Controls upper two bits in the octet
tagClassBit :: TagClass -> Word8
tagClassBit = \case
  Universal -> 0
  Application -> 64
  ContextSpecific -> 128
  Private -> 192

sequenceTag :: Tag
sequenceTag = Tag Universal 16

tagNumStringType :: StringType -> Int
tagNumStringType = \case
  Utf8String -> 12
  NumericString -> 18
  PrintableString -> 19
  TeletexString -> 20
  VideotexString -> 21
  IA5String -> 22
  GraphicString -> 25
  VisibleString -> 26
  GeneralString -> 27
  UniversalString -> 28
  CharacterString -> 29
  BmpString -> 30
